src/lightning/fabric/plugins/environments/slurm.py

```diff
@@ -141,6 +141,22 @@ def local_rank(self) -> int:
     def node_rank(self) -> int:
         return int(os.environ["SLURM_NODEID"])
 
+    def validate_settings(self, num_devices: int, num_nodes: int) -> None:
+        if _is_slurm_interactive_mode():
+            return
+        ntasks_per_node = os.environ.get("SLURM_NTASKS_PER_NODE")
+        if ntasks_per_node is not None and int(ntasks_per_node) != num_devices:
+            raise ValueError(
+                f"You set `devices={num_devices}` in Lightning, but the number of tasks per node configured in SLURM"
+                f" `--ntasks-per-node={ntasks_per_node}` does not match. HINT: Set `devices={ntasks_per_node}`."
+            )
+        nnodes = os.environ.get("SLURM_NNODES")
+        if nnodes is not None and int(nnodes) != num_nodes:
```

Comment:
hi @awaelchli, just a minor note. I came across this error when using lightning nightly build
```
ValueError: You set `num_nodes=4` in Lightning, but the number of nodes configured in SLURM `--nodes=4` does not match. HINT: Set `num_nodes=4`.
```

I wasn't able to investigate this deeply but wanted to flag it here in case its of help to you. Not sure if it happens due to type of num_nodes, but casting it to int gets the code through.

Reply:
Hey @Sumith1896 
That's very odd. I don't understand how the check above can pass 4 != 4 when the raised error message shows that both nnodes=4 and num_nodes=4. 

`int(nnodes)` makes it an int and `num_nodes` should already be one.

> but casting it to int gets the code through.

Could you tell me where a cast to int is needed? Is `num_nodes` not an int for some reason?

Reply:
`num_nodes` wasn't an int in my case. I'm quite puzzled myself as the same lightning env worked on different compute instances/slurm fine. Not sure if its a slurm version / instance specific. This same PR seems to have removed cast of num_nodes to int, maybe that's the case. I can also help narrow it down for you on my end if needed.

Reply:
Are you using Trainer or Fabric?
I removed the cast because I deduced it as redundent, as earlier in the code the num_nodes is assumed to be int and converted to int. In Fabric, this happens at the very beginning here: https://github.com/Lightning-AI/lightning/blob/32415c68ae981a83c72d9587cb02769a4a5b894e/src/lightning/fabric/connector.py#L116
And in Trainer, the num_nodes input is expected to be int from the start: https://github.com/Lightning-AI/lightning/blob/32415c68ae981a83c72d9587cb02769a4a5b894e/src/lightning/pytorch/trainer/connectors/accelerator_connector.py#L81

Random guess. Do you have an argparser that takes the num_nodes argument from the command line, and keeps it as string instead of converting to int?

Reply:
Could you do me a favor and on your system check where in the connector I linked above the num_nodes is not an int. Is it a the input or later?

Reply:
@awaelchli ah yes! It's a faulty argparse that sends `num_nodes` as `str`, I'm a bit surprised by the non-deterministic behavior, but it doesn't seem to be lightning's fault. One could also check if `num_nodes` is an `str` that could be cast as `int` but I'll let you take the call. Thanks for engaging!

Reply:
Thanks for confirming! I added input validation yesterday: https://github.com/Lightning-AI/lightning/pull/18598, this should help resolve any confusions in the future.